src/proto/batch/mod.rs

```diff
@@ -0,0 +1,386 @@
+/// let child_batch = Batch::builder()
+///     .description("Child batch description")
+///     .with_success_callback(child_cb);
+///
+/// let mut parent = prod.start_batch(parent_batch)?;
+/// parent.add(parent_job1)?;
+/// parent.add(parent_job2)?;
+/// let mut child = parent.start_batch(child_batch)?;
+/// child.add(child_job1)?;
+/// child.add(child_job2)?;
+///
+/// child.commit()?;
+/// parent.commit()?;
+///
+/// # Ok::<(), Error>(())
+/// ```
+///
+/// In the example above, there is a single level nesting, but you can nest those batches as deep as you wish,
+/// effectively building a pipeline this way, since the Faktory guarantees that callback jobs will not be queued unless
+/// the batch gets committed.
+///
+/// You can retieve the batch status using a [`Tracker`](struct.Tracker.html):
+/// ```no_run
+/// # use faktory::Error;
+/// # use faktory::{Producer, Job, Batch, Tracker};
+/// let mut prod = Producer::connect(None)?;
+/// let job = Job::builder("job_type").build();
+/// let cb_job = Job::builder("callback_job_type").build();
+/// let b = Batch::builder()
+///     .description("Batch description")
+///     .with_complete_callback(cb_job);
+///
+/// let mut b = prod.start_batch(b)?;
+/// let bid = b.id().to_string();
+/// b.add(job)?;
+/// b.commit()?;
+///
+/// let mut t = Tracker::connect(None)?;
+/// let s = t.get_batch_status(bid)?.unwrap();
+/// assert_eq!(s.total, 1);
+/// assert_eq!(s.pending, 1);
+/// assert_eq!(s.description, Some("Batch description".into()));
+/// assert_eq!(s.complete_callback_state, ""); // has not been queued;
+/// # Ok::<(), Error>(())
+/// ```
+#[derive(Builder, Debug, Serialize)]
+#[builder(
+    custom_constructor,
+    pattern = "owned",
+    setter(into),
+    build_fn(name = "try_build", private)
+)]
+pub struct Batch {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[builder(setter(skip))]
+    parent_bid: Option<String>,
+
+    /// Batch description for Faktory WEB UI.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[builder(setter(custom), default = "None")]
+    pub description: Option<String>,
+
+    /// On success callback.
+    ///
+    /// This job will be queued by the Faktory server provided
+    /// all the jobs belonging to this batch have been executed successfully.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[builder(setter(skip))]
+    pub(crate) success: Option<Job>,
+
+    /// On complete callback.
+    ///
+    /// This job will be queue by the Faktory server after all the jobs
+    /// belonging to this batch have been executed, even if one/some/all
+    /// of the workers have failed.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[builder(setter(skip))]
+    pub(crate) complete: Option<Job>,
+}
+
+impl Batch {
+    /// Create a new `BatchBuilder`.
+    pub fn builder() -> BatchBuilder {
+        BatchBuilder::new()
+    }
+}
+
+impl BatchBuilder {
+    fn build(self) -> Batch {
```

Comment:
As currently written, it is not possible to construct a batch that has neither a success or a complete callback. Is that intentional?

Reply:
It is. The batch should have either/both callbacks as per [docs](https://github.com/contribsys/faktory/wiki/Ent-Batches#batch-new-json). 

Interestingly, the Go client will not force the user to specify either/both of them, committing such a batch will make the server error back with:
```
ERR Batch must define a success or complete callback
```

I learned it accidentally: I just forgot to do so initially when writing this [test](https://github.com/rustworthy/faktory/actions/runs/7765783083/job/21180846057)